compiler/src/dotty/tools/dotc/transform/patmat/Space.scala

```diff
@@ -664,31 +751,35 @@ class SpaceEngine(implicit ctx: Context) extends SpaceLogic {
 
     val patternSpace = cases.map({ x =>
       val space = project(x.pat)
-      debug.println(s"${x.pat.show} projects to ${show(space)}")
+      debug.println(s"${x.pat.show} ====> ${show(space)}")
       space
     }).reduce((a, b) => Or(List(a, b)))
-    val uncovered = simplify(minus(Typ(selTyp, true), patternSpace))
+    val uncovered = simplify(minus(Typ(selTyp, true), patternSpace), aggressive = true)
 
     if (uncovered != Empty)
-      ctx.warning(PatternMatchExhaustivity(show(uncovered)), _match.pos)
+      ctx.warning(PatternMatchExhaustivity(show(uncovered)), sel.pos)
   }
 
   def checkRedundancy(_match: Match): Unit = {
     val Match(sel, cases) = _match
     // ignore selector type for now
     // val selTyp = sel.tpe.widen.deAnonymize.dealias
 
+    if (cases.length == 1) return
+
     // starts from the second, the first can't be redundant
     (1 until cases.length).foreach { i =>
-      // in redundancy check, take guard as false, take extractor as match
-      // nothing in order to soundly approximate
+      // in redundancy check, take guard as false in order to soundly approximate
       val prevs = cases.take(i).map { x =>
-        if (x.guard.isEmpty) project(x.pat, false)
+        if (x.guard.isEmpty) project(x.pat)
         else Empty
       }.reduce((a, b) => Or(List(a, b)))
 
       val curr = project(cases(i).pat)
 
+      debug.println(s"---------------reachable? ${show(curr)}")
```

Comment:
I suggest adding a `Printer` for the exhaustiveness checker in https://github.com/lampepfl/dotty/blob/master/compiler/src/dotty/tools/dotc/config/Printers.scala instead of using `debug` here.